tensorflow_encrypted/tensor/native.py

```diff
@@ -0,0 +1,215 @@
+from __future__ import absolute_import
+
+import numpy as np
+import tensorflow as tf
+from typing import Union, Optional, List, Dict, Any, Tuple
+
+from ..config import run
+
+INT_TYPE = tf.int32
+
+
+class NativeTensor(object):
+    int_type = INT_TYPE
+
+    def __init__(self, value: Union[np.ndarray, tf.Tensor], modulus: int) -> None:
+        self.value = value
+        self.modulus = modulus
+
+    @staticmethod
+    def from_native(value: Union[np.ndarray, tf.Tensor], modulus: int) -> 'NativeTensor':
+        assert isinstance(value, (np.ndarray, tf.Tensor)), type(value)
+        return NativeTensor(value, modulus)
+
+    @staticmethod
+    def sample_uniform(shape: Union[Tuple[int, ...], tf.TensorShape], modulus: int) -> 'NativeTensor':
+        return NativeTensor(tf.random_uniform(shape=shape, dtype=INT_TYPE, maxval=modulus), modulus)
+
+    @staticmethod
+    def sample_bounded(shape: List[int], bitlength: int) -> 'NativeTensor':
+        raise NotImplementedError()
+
+    def eval(self, sess: tf.Session, feed_dict: Dict[Any, Any]={},
+             tag: Optional[str]=None) -> 'NativeTensor':
+        return NativeTensor(run(sess, self.value, feed_dict=feed_dict, tag=tag), self.modulus)
+
+    def __repr__(self) -> str:
+        return 'NativeTensor({})'.format(self.shape)
+
+    @property
+    def shape(self) -> Union[Tuple[int, ...], tf.TensorShape]:
+        if self.value is None:
+            raise Exception("Can't call 'shape' on a empty tensor")
+
+        return self.value.shape
+
+    def __add__(self, other: Union['NativeTensor', int]) -> 'NativeTensor':
+        return self.add(other)
+
+    def __sub__(self, other: Union['NativeTensor', int]) -> 'NativeTensor':
+        return self.sub(other)
+
+    def __mul__(self, other: Union['NativeTensor', int]) -> 'NativeTensor':
+        return self.mul(other)
+
+    def __mod__(self, k: Union['NativeTensor', int]) -> 'NativeTensor':
+        return self.mod(k)
+
+    def add(self, other: Union['NativeTensor', int]) -> 'NativeTensor':
+        x, y = _lift(self, self.modulus), _lift(other, self.modulus)
+        return NativeTensor(x.value + y.value % self.modulus, self.modulus)
```

Comment:
`%` binds strong than `+`, so you have to do `(x.value + y.value) % self.modulus`; same for `-`

Reply:
Ahh yes, good call!